### DISCUSSION TITLE
Exclude unsuccessful inputs from ConsoleHost_history.txt

### DISCUSSION THREAD

**Discussion by God-damnit-all** [lzybkr/PSReadLine#3760] [Q&A]
Sometimes I look through ConsoleHost_history.txt to figure out how or why I did something prior. _(I take notes but sometimes I'm in a rush to do something, you know how it is.)_

However, I really wish that only successful inputs would make it to the log, i.e. if an error was thrown as a result of the input, don't log it. Is there a way to accomplish this? I don't mind if it impacts the performance within the terminal as long as it doesn't impact the performance of scripts.

**Comment by daxian-dbw** [comment]
There is no good way of doing this because PSReadLine save the command line to history file before returning it to PowerShell engine to execute.

There is a way to validate the command line before returning it to PowerShell engine, it's done the combination of `Set-PSReadLineOption -CommandValidationHandler` and bind <kbd>Enter</kbd> key with `ValidateAndAcceptLine` function (an [example](https://github.com/PowerShell/PSReadLine/blob/master/PSReadLine/SamplePSReadLineProfile.ps1#L567)). However, it won't capture all the possible failures as some failures can only be found when actually executing the command line.

**Reply by God-damnit-all** [reply]
> There is no good way of doing this because PSReadLine save the command line to history file before returning it to PowerShell engine to execute.
> 
> There is a way to validate the command line before returning it to PowerShell engine, it's done the combination of `Set-PSReadLineOption -CommandValidationHandler` and bind Enter key with `ValidateAndAcceptLine` function (an [example](https://github.com/PowerShell/PSReadLine/blob/master/PSReadLine/SamplePSReadLineProfile.ps1#L567)). However, it won't capture all the possible failures as some failures can only be found when actually executing the command line.

Well, let's start there. I notice that this places the error below the console. I'd ideally to make pressing Enter do the following:

1. Programmatically determine if the current input passes validation
a. If yes, behave as normal
b. If no, proceed to Step 2
2. Write-Error or throw the validation exception to the console
3. Clear the current input buffer
4. Add the input to the session console history but **not** ConsoleHost_history.txt

I know Step 3 is doable via `[Microsoft.PowerShell.PSConsoleReadLine]::RevertLine()`, but I have no idea about the rest. I tried putting `[Microsoft.PowerShell.PSConsoleReadLine]::ValidateAndAcceptLine()` in a try/catch block and it didn't appear to trigger the catch block at all. Is there another function I can use for validating that will trigger a catch block?

As for Step 4, I know session history is a thing. If nothing else, I could have it trigger `Set-PSReadlineOption -HistorySaveStyle SaveNothing` before running `[Microsoft.PowerShell.PSConsoleReadLine]::AcceptLine()` and then revert it back to its previous setting.
